smarts/core/opendrive_road_network.py

```diff
@@ -0,0 +1,1810 @@
+
+            # right_edge
+            right_edge_shape = self._lane_polygon[
+                reference_line_vertices_len : len(self._lane_polygon) - 1
+            ]
+            right_offset = offset_along_shape(point[:2], right_edge_shape)
+            x, y = position_at_shape_offset(right_edge_shape, right_offset)
+            right_edge = Point(x, y)
+            return left_edge, right_edge
+
+        @lru_cache(8)
+        def vector_at_offset(self, start_offset: float) -> np.ndarray:
+            return super().vector_at_offset(start_offset)
+
+        @lru_cache(maxsize=8)
+        def center_pose_at_point(self, point: Point) -> Pose:
+            return super().center_pose_at_point(point)
+
+        @lru_cache(maxsize=8)
+        def curvature_radius_at_offset(
+            self, offset: float, lookahead: int = 5
+        ) -> float:
+            return super().curvature_radius_at_offset(offset, lookahead)
+
+        @lru_cache(maxsize=8)
+        def width_at_offset(self, lane_point_s: float) -> float:
+            if self._lane_elem_index < 0:
+                road_offset = lane_point_s + self.road.s_pos
+            else:
+                road_offset = (self._length - lane_point_s) + self.road.s_pos
+            inner_boundary, outer_boundary = self._lane_boundaries
+            t_outer = outer_boundary.calc_t(
+                road_offset, self.road.s_pos, self._lane_elem_index
+            )
+            t_inner = inner_boundary.calc_t(
+                road_offset, self.road.s_pos, self._lane_elem_index
+            )
+            return abs(t_outer - t_inner)
+
+        @lru_cache(maxsize=4)
+        def shape(self, width: float = 0.0, buffer_width: float = 0.0) -> Polygon:
+            if buffer_width == 0.0:
+                return Polygon(self._lane_polygon)
+            buffered_polygon = self.compute_lane_polygon(buffer_width / 2)
+            return Polygon(buffered_polygon)
+
+    def lane_by_id(self, lane_id: str) -> RoadMap.Lane:
+        lane = self._lanes.get(lane_id)
+        if not lane:
+            self._log.warning(
+                f"OpenDriveRoadNetwork got request for unknown lane_id '{lane_id}'"
+            )
+        return lane
+
+    class Road(RoadMap.Road, Surface):
+        def __init__(
+            self,
+            road_id: str,
+            is_junction: bool,
+            length: float,
+            s_pos: float,
+            total_lanes: int,
+        ):
+            super().__init__(road_id)
+            self._log = logging.getLogger(self.__class__.__name__)
+            self._road_id = road_id
+            self._is_junction = is_junction
+            self._length = length
+            self._s_pos = s_pos
+            self._is_drivable = False
+            self._lanes = []
+            self._bounding_box = []
+            self._incoming_roads = []
+            self._outgoing_roads = []
+            self._parallel_roads = []
+            self._total_lanes = total_lanes
```

Comment:
Yes, but we need the value when we are parsing the `.xodr` file for the first time and are creating the `Lane` objects for a given `Road` object. So we can't access the total_lanes from the `len(self._lanes)` since we haven't parsed all the lanes in that `Road`. So we try to access that information from the length of the `lanes `list of the `RoadElement` in `opendriveparser` object.